### DISCUSSION TITLE
Add Field Grouping

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#3526] [Feature Requests]
This would allow grouping fields (beyond just dividers), and eventually paves the way for conditional groups.

**Comment by davidbwaters** [comment]
IMHO this should be high priority in the age of components and seems like it could pretty much be added by allowing fields in a many-to-one and allow 2 - 3 levels of nesting within.  

I like the way Strapi has components (field groups) and 'dynamic zones' (sortable lists where a selection of allowed components can be added).  It's very flexible and basically changes the experience from filling in a spreadsheet to a flexible 'layout builder' would allow directus to be used to create complex content layouts without messing with the content model.  

Strapi stores all instances of a component in a one table per component and they are added to the 'zone' or parent component as references.

**Reply by davidbwaters** [reply]
![79272475-e2d46600-7e77-11ea-8e1d-54261d224131](https://user-images.githubusercontent.com/5247627/106410620-403ea100-6411-11eb-8e6b-a27c3b00a7eb.png)

**Reply by philleepflorence** [reply]
This really looks cool if the primary use of Directus is for the development of websites.

I wonder how we could we imagine this concept if Directus is used for Websites, Web Applications, API Applications, Data Management, and more?

Pages on Web Applications may not have the same structure as Blogs or Marketing Websites.

**Reply by davidbwaters** [reply]
I think this could be useful beyond websites to anything that needs a 'rich' layout, so forms etc., which most of those use.  For some of those, just managing DB tables it enough, which Directus does great already.  As long as it doesn't hurt the experience for others, I would think it would be welcome and there are core features like a publishing system don't fit all use cases in the core.   

From a data perspective IMO, now we have the choice of a singleton or adding to a single monolithic list of collections.  Adding the ability to structure collections of smaller collections definitely has a use case beyond web layouts.  

For example, a reference to a dropdown or group of checkboxes would enable categories, tags, taxonomies, etc without having to add tags in the same list alongside Users.

But if this was added as part of an extension, that would also be welcome.  I would bet the node rewrite made things more modular so that could be an option.

**Reply by benhaynes** [reply]
Hey @davidbwaters — thank you for posting this! Here are my thoughts:

Yes, what you're describing is a really powerful feature, and one that we've talked about internally for a while. Basically you would be able to have a M2O interface show its [**form** on the parent](https://github.com/directus/directus/discussions/3474) instead of opening in a drawer. The same would ply to a O2M, but this would essentially be the Relational Repeater interface. Both of these are good ideas (and I would love to fast-track them)... but they are different from _this_ ticket.

This ticket is about grouping the parent fields on a form. So if you have a collection with a lot of fields, you can organize them a bit better. It goes a step further to allow potentially collapsing groups, and using the groups in "conditionals". So that you can say: "Show this Group of fields once X has been satisfied".

In any event, both of these are important... but you might want to add you thoughts to that #3474 ticket... since that one is a bit faster to implement too. :)

**Reply by rijkvanzanten** [reply]
(to quickly clarify: 

> Strapi stores all instances of a component in a one table per component and they are added to the 'zone' or parent component as references.

This is the same as our many-to-any relationship (w/ the builder interface) 👍🏻 )

**Reply by benhaynes** [reply]
Correct. And we _could_ have a M2A interface that renders each related form inline... but again, that is more of a conversation for #3474

**Reply by davidbwaters** [reply]
Thanks.  I guess I like the abstraction of Strapi's approach kinda merges these functionalities with what are essentially field group templates available in different collections. as relationships.

**Reply by benhaynes** [reply]
Yeah! That's exactly what #3474 is ... you can define a field group as a collection of fields, and then inject that (relationally) into any other collection.

**Reply by JLKM** [reply]
A O2M interface on a form showing related records (with the possibility to add more) is a big wish of mine. Could work like as an embedded list with its own template - filtered on the related parent record. It's just about the only thing I miss on this absolutely wonderful project.

**Reply by benhaynes** [reply]
@JLKM — could you describe this a bit more? It sounds like the O2M Interface we have now, which lists items and clicking opens their form in a drawer. What differences are you looking for? Something to do with the relational _forms_ bing inline on the page?

**Reply by JLKM** [reply]
@benhaynes Sorry. My bad. Directus 9 is even more than capable of what I wished for. I got it working by using the "Create Field in Advanced Mode" and create the relations from the M2O-side - including the corresponding O2M-field. It works like a charm. Really impressive with all that templating + fetching fields across tables. Pretty mindblowing, what has to be going on behind the scenes here. Nothing now prevents me from going ahead with my project. Thanks again for this awesome project.

**Reply by benhaynes** [reply]
Haha, nice! That's great to hear. I'm always interested in how we can improve the Core Interfaces... so I was curious!

**Comment by rijkvanzanten** [comment]
This has since been added!
